skimage/registration/_optical_flow.py

```diff
@@ -216,3 +216,164 @@ def optical_flow_tvl1(reference_image, moving_image,
                      tol=tol, prefilter=prefilter)
 
     return coarse_to_fine(reference_image, moving_image, solver, dtype=dtype)
+
+
+def _ilk(reference_image, moving_image, flow0, radius, num_warp, gaussian,
+         prefilter):
+    """Iterative Lucas-Kanade (iLK) solver for optical flow estimation.
+
+    Parameters
+    ----------
+    reference_image : ndarray, shape (M, N[, P[, ...]])
+        The first gray scale image of the sequence.
+    moving_image : ndarray, shape (M, N[, P[, ...]])
+        The second gray scale image of the sequence.
+    flow0 : ndarray, shape (reference_image.ndim, M, N[, P[, ...]])
+        Initialization for the vector field.
+    radius : int
+        Radius of the window considered around each pixel.
+    num_warp : int
+        Number of times moving_image is warped.
+    gaussian : bool
+        if True, a gaussian kernel is used for the local
+        intagration. Otherwise, a uniform kernel is used.
+    prefilter : bool
+        whether to prefilter the estimated optical flow before each
+        image warp. This helps to remove the potential outliers.
+
+    Returns
+    -------
+    flow : ndarray, shape ((reference_image.ndim, M, N[, P[, ...]])
+        The estimated optical flow components for each axis.
+
+    """
+    dtype = reference_image.dtype
+    ndim = reference_image.ndim
+    size = 2 * radius + 1
+
+    if gaussian:
+        sigma = (0, ) + ndim * (size / 4, )
+        filter_func = partial(ndi.gaussian_filter, sigma=sigma, mode='mirror')
+    else:
+        filter_func = partial(ndi.uniform_filter, size=(1, ) + ndim * (size, ),
+                              mode='mirror')
+
+    flow = flow0
+    coef = np.zeros((int((ndim * (ndim + 1)) / 2 + ndim), )
+                    + reference_image.shape, dtype=dtype)
+    # For each pixel location (i, j), the optical flow X = flow[:, i, j]
+    # is the solution of the ndim x ndim linear system
+    # A[i, j] * X = b[i, j]
+    A = np.zeros(reference_image.shape + (ndim, ndim), dtype=dtype)
+    b = np.zeros(reference_image.shape + (ndim, ), dtype=dtype)
+
+    grid = np.meshgrid(*[np.arange(n, dtype=dtype)
+                         for n in reference_image.shape], indexing='ij')
```

Comment:
Could you add a comment here and in the function docstring that the space required for this computation is (as far as I can tell) ndim * (ndim + 2) of the original image size, in addition to the image itself?

Reply:
Actually, that's just from the above three lines, there's lots of other volumes being created willy-nilly... LOL

Reply:
You don't have to store the full dense grid if you use
```Python
grid = np.meshgrid(*[np.arange(n, dtype=float) for n in (256, 256, 192)], indexing='ij', sparse=True)
```
It appears `grid` is only ever used in an addition (`grid + flow`) below. In the sparse grid case, an equivalent addition would be:
```Python
np.stack([g + f for g, f in zip(grid, flow)], axis=0)
```
which is a bit less readable, but does allow not storing the full grid. A more verbose, sparse grid addition helper would be:
```Python
def _add_flow(grid, flow):
    out = flow.copy()
    for n, g in enumerate(grid):
        out[n, ...] += g
    return out
```

Reply:
Thank you @grlee77 for this suggestion, I vote for defining a helper function which can also be used in `_tvl1` :wink:.